src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+    ty' <- occ accessOnce ty
+    return $ DeclBinding ann ty' expr'
+
+instance HasOCC Expr where
+  occ a expr = case expr of
+    (TabApp array (NE.toList -> ixs)) -> do
+      (a', ixs') <- go a ixs
+      array' <- occ a' array
+      return $ TabApp array' $ NE.fromList ixs'
+    (Case scrut alts ty effs) -> do
+      scrut' <- occ accessOnce scrut
+      scrutIx <- summary scrut
+      (alts', innerFVs) <- unzip <$> mapM (isolated . occAlt a scrutIx) alts
+      modify (<> foldl' Occ.max zero innerFVs)
+      -- TODO(optimization) Is free variable counting enough for types?
+      ty' <- occ accessOnce ty
+      countFreeVarsAsOccurrences effs
+      return $ Case scrut' alts' ty' effs
+    _ -> generic
+    where
+      generic = toE <$> (occ a $ fromE expr)
+      go acc [] = return (acc, [])
+      go acc (ix:ixs) = do
+        (acc', ixs') <- go acc ixs
+        (summ, ix') <- occurrenceAndSummary ix
+        return (location summ acc', ix':ixs')
+
+-- Arguments: Usage of the return value, summary of the scrutinee, the
+-- alternative itself.
+occAlt :: Access n -> IxExpr n -> Alt n -> OCCM n (Alt n)
+occAlt acc scrut alt = do
+  refreshAbs alt \b@(nb:>_) body -> do
+    scrutIx <- unknown $ sink scrut
+    extend nb scrutIx do
+      body' <- occ (sink acc) body
+      return $ Abs b body'
+
+occurrenceAndSummary :: Atom n -> OCCM n (IxExpr n, Atom n)
+occurrenceAndSummary atom = do
+  atom' <- occ accessOnce atom
+  ix <- summary atom'
+  return (ix, atom')
+
+instance HasOCC (ComposeE PrimHof Atom) where
+  occ a (ComposeE hof) = ComposeE <$> case hof of
+    For ann ixDict bd -> do
+      ixDict' <- inlinedLater ixDict
+      bd' <- occLam bd \b body -> do
+        extend b (Occ.Var $ binderName b) do
+          (body', bodyFV) <- isolated (occ accessOnce body)
+          modify (<> abstractFor b bodyFV)
+          return body'
+      return $ For ann ixDict' bd'
+    While bd -> While <$> occLam bd \b' body -> do
+      -- I think IxAll is right, since indexing with a value of Unit type does
+      -- mean indexing with every possible value of the index type.
+      extend b' IxAll do
+        (body', bodyFV) <- isolated (occ accessOnce body)
+        modify (<> useManyTimes bodyFV)
+        return body'
+    RunReader ini bd -> do
+      ini' <- occ accessOnce ini
+      iniIx <- summary ini
+      bd' <- oneShot2 a (Deterministic []) iniIx bd
+      return $ RunReader ini' bd'
+    RunWriter Nothing (BaseMonoid empty combine) bd -> do
+      -- We will process the combining function when we meet it in MExtend ops
+      -- (but we won't attempt to eliminate dead code in it).
+      empty' <- occ accessOnce empty
+      -- The body is exactly one-shot, and shouldn't index with the reference
+      -- parameter.  (But if it does, Deterministic [] seems reasonable, because
+      -- RunWriter syntehsizes the reference, presumably independently of
+      -- anything else.)  Same for the heap parameter.
```

Comment:
Could we insert `undefined` in place of the `Deterministic`? While I would be surprised if anybody ever managed to index with a reference, I feel a little uneasy just dumping garbage values in the map.

Reply:
This thunk is actually forced, so no.  Wrote out a more complete comment defending `Deterministic []` as sound here, including an example of why `undefined` should not be expected to work.